clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java

```diff
@@ -318,4 +318,11 @@ public AttributeList setAttributes(AttributeList list) {
                                       + ".(whitelist/blacklist) is not a valid regular expression");
         }
     }
+
+    @Override
+    public synchronized void contextChange(MetricsContext metricsContext) {
+        if (this.prefix == null || this.prefix.isEmpty()) {
+            this.prefix = metricsContext.metadata().get(MetricsContext.NAMESPACE);
```

Comment:
we need a safety mechanism to ensure the prefix cannot be changed after the first metric has been registered. This doesn't prevent this. One way to ensure this would be to check that the mbeans map is empty and throw an IllegalStateException if the prefix is changed when the map is not empty.